Rework ERD and Entity Definitions (#2)
seenons/seenons-be-assignment: README.md

able to work in this ecosystem. Your own experience will be taken into account when reviewing your assignment, so no
need to panic if you are not familiar with them.

You can complete the assignment using the following ecosystems:

- Node JS/TS
- .NET (Core) 5+
- Java / Kotlin

> - Python

Reviewer: 😞 😢 😱

Reviewer 2: ODIN is written in Python, so I agree with Caio that we should allow the language in the assignment as well

Reviewer 3: I actually completely missed that part when I first reviewed it.
The original idea we had with Lukas was that a we need people who can adapt to a new eco system. Since we're using Typescript internally, we wanted to force the assignment to be in TS. Hence having a skeleton, to reduce the overhead of having to setup everything if you're not familiar with it.

In this light I would either allow any language (why can't they do it in C++ if they want to?) or only allow TS (and maybe allow any language for more Junior positions, so they are more comfortable).

Author: If this is something we don't want, then I can remove it, but maybe lets discuss it or rewrite it a bit (where we state that we'd rather you do it in JS/TS).

To me, ideally the candidate should deliver using the assignment using the base to become familiar with the stack, this should also help us have a nice baseline to evaluate how things were done. 

Also, since "time" is precious during the hiring process, sometimes the candidate won't continue due to the fact that they'd rather do it in a stack they are most familiar with to be able to continue multiple processes, and by not allowing them to do it this way we risk losing a potential candidate that has too many processes going on (it happened already last week).
Sometimes their expertise doesn't transition immediately, but the engineering principles are the same

Now for the C/C++ part, do we have people that can and are willing to evaluate?

I can evaluate in JS/TS, Java, Kotlin, C# while others can also evaluate in Python (this was included because we are also looking for a new joiner for `spops`, which would need some snakezzz language knowledge

Reviewer 3: Yeah that's the other side of the coin. Maybe at least stating that TS is preferred as this is what we use, and that we would take it into account if they have no or little prior knowledge of the eco-system, so they should not hesitate to give it a try if they feel like it.
I'm a bit rusty (uhuh) but I can probably still evaluate stuff in C++. Also in Python btw.

Author: Done and added a little disclaimer to the langgages/ecosystems/stacks

## Data Model Representation

### Waste Stream

A Waste Stream can be described as:
A specific type of waste that can be collected by a Service Provider and is registered for collection by a Customer.

| id | label |
|:--:|:-----:|
| 1  | paper |
| 2  | metal |
| 3  | glass |

### Service Provider

Can be described as:
A business that can collect a specific set of waste streams.

| id |      name      |                 address                 | coverages |
|:--:|:--------------:|:---------------------------------------:|:---------:|
| 1  |    Unwasted    |   Stationplein, 1, 1012 AB Amsterdam    |  [1, 2]   |
| 2  | Bluecollection | Prins Hendrikkade, 1, 1012 JD Amsterdam |    [3]    |

We have two service providers, "Unwasted" and "Bluecollection".

### Service Provider Coverage

Can be described as:

> The waste streams in which a service provider has covered in a postal code range, for a given set of weekdays.

Reviewer 3: ```suggestion
The waste streams a service provider can cover in a postal code range, for a given set of weekdays.
```
or at least there is a typo ;)

#### Opportunities

- Enhance testing capabilities and re-usability.
- Implement a Database Provider.
- Implement caching strategy.
- Can you apply a better design pattern for the overall registration and data modeling?

> - Show us your understanding of Domain Driven Design and Loose Coupling?

Reviewer 3: If you want to make it a question, then really make it a question. :thinking: 
This just sounds like you're kindly asking them to please do so if really they don't mind. :smile:
